app/scripts/services/environment.js

```diff
@@ -0,0 +1,77 @@
+'use strict';
+
+angular.module("openshiftConsole")
+  .factory("EnvironmentService",
+           function($filter,
+                    keyValueEditorUtils) {
+    var altTextForValueFrom = $filter('altTextForValueFrom');
+    var getContainers = function(set) {
+      return _.get(set, 'spec.template.spec.containers', []);
+    };
+
+    return {
+      // Make sure there is an `env` property for each container and add in alt
+      // text for any value from entries.
+      // Note: This modifies object. It should only be called on a copy.
+      normalizeEnvironment: function(object) {
+        var containers = getContainers(object);
+        _.each(containers, function(container) {
+          container.env = container.env || [];
+          // check valueFrom attribs and set an alt text for display if present
+          _.each(container.env, altTextForValueFrom);
+        });
+      },
+
+      // Call `keyValueEditorUtils.compactEntries` on the env for each container.
+      // Note: This modifies object. It should only be called on a copy.
+      compactEnvironment: function(object) {
+        var containers = getContainers(object);
+        _.each(containers, function(container) {
+          container.env = keyValueEditorUtils.compactEntries(container.env);
+        });
+      },
+
+      // Copy and normalize the environment for editing using the key value editor.
+      copy: function(object) {
+        var copy = angular.copy(object);
+        this.normalizeEnvironment(copy);
+        return copy;
+      },
+
+      // Compare the current and previous versions of an object to see if any
+      // of the environment variables have changed.
+      isEnvironmentEqual: function(left, right) {
+        var leftContainers = getContainers(left);
+        var rightContainers = getContainers(right);
+        if (leftContainers.length !== rightContainers.length) {
+          return false;
+        }
+
+        var i, leftEnv, rightEnv;
+        for (i = 0; i < leftContainers.length; i++) {
+          leftEnv = leftContainers[i].env || [];
```

Comment:
i feel like we should be validating that the container names are actually the same, what if i took out a container, added a brand new container, the length of containers array is the same but they are no longer equal. Especially since merge edits below assumes containers in the same positions are in fact the same.